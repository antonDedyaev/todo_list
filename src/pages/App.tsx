import getFilteredTasks from 'helpers/getFilteredTasks';
import TaskList from '../components/TaskList';
import React, { useState, type KeyboardEvent, type MouseEvent, type FC } from 'react';
import getSavedTasks from 'helpers/getSavedTasks';

const App: FC = () => {
  const allTasks = getSavedTasks();
  const [inputValue, setInputValue] = useState('');
  const [filteredList, setFilteredList] = useState(allTasks);
  const [activeTasksLeft, setActiveTasksLeft] = useState(allTasks.filter(task => !task.isCompleted));

  const handleSaveInput = (e: KeyboardEvent<HTMLInputElement>): void => {
    if (e.key === 'Enter') {
      const currentValue = (e.target as HTMLInputElement).value;
      const newTask = { id: Date.now(), text: currentValue, isCompleted: false };
      allTasks.push(newTask);
      localStorage.setItem('tasks', JSON.stringify(allTasks));
      setFilteredList(allTasks);
      setActiveTasksLeft(allTasks.filter(task => !task.isCompleted));
      setInputValue('');
    }
  };

  const handleSelectFilter = (e: MouseEvent<HTMLDivElement>): void => {
    document.querySelector('.container__filterButton_active')?.classList.remove('container__filterButton_active');
    const target = e.target as HTMLButtonElement;
    target.classList.add('container__filterButton_active');
    if (target.textContent !== null) {
      const filteredTasks = getFilteredTasks(target.textContent, getSavedTasks());
      setFilteredList(filteredTasks);
    }
  };

  const handleClearCompletedTasks = (): void => {
    localStorage.setItem('tasks', JSON.stringify(activeTasksLeft));
    setFilteredList(activeTasksLeft);
  };

  return (
    <div className="container">
      <div className="container__content">
        <header className="container__contentHeader">
          <h1>todos</h1>
        </header>
        <section className="container__contentBody">
          <div className="container__inputLine">
            <input
              type="text"
              value={inputValue}
              placeholder="What needs to be done?"
              data-testid="main-input"
              onChange={e => {
                setInputValue(e.target.value);
              }}
              onKeyDown={e => {
                handleSaveInput(e);
              }}
            />
          </div>
          <div className="container__list">
            <TaskList tasks={filteredList} setActiveTasksLeft={setActiveTasksLeft} />
          </div>
        </section>
        <footer className="container__contentFooter">
          <span>{`${activeTasksLeft.length} ${activeTasksLeft.length === 1 ? 'item' : 'items'} left`}</span>
          <div
            className="container__listFilters"
            onClick={e => {
              handleSelectFilter(e);
            }}>
            <button className="container__filterButton container__filterButton_active" data-testid="display-all">
              All
            </button>
            <button className="container__filterButton" data-testid="active-todos">
              Active
            </button>
            <button className="container__filterButton">Completed</button>
          </div>
          <button onClick={handleClearCompletedTasks} data-testid="clear-all">
            Clear completed
          </button>
        </footer>
        <div className="container__backPage container__backPage_first"></div>
        <div className="container__backPage container__backPage_second"></div>
      </div>
    </div>
  );
};

export default App;
