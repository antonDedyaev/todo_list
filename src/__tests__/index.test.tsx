import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from 'pages/App';
import React from 'react';

describe('CORRECT RENDERING', () => {
  test('renders input', () => {
    render(<App />);
    const inputElement = screen.getByPlaceholderText(/what needs to be done?/i);
    expect(inputElement).toBeInTheDocument();
  });

  test('default filter selected', () => {
    render(<App />);
    const buttonElement = screen.getByTestId('display-all');
    expect(buttonElement).toHaveClass('container__filterButton_active');
  });
});

describe('MAIN FUNCTIONS', () => {
  test('adds a new todo', () => {
    render(<App />);

    expect(screen.queryByTestId('list-item')).not.toBeInTheDocument();
    const inputElement = screen.getByPlaceholderText(/what needs to be done?/i);

    userEvent.type(inputElement, 'Test');
    expect(inputElement).toHaveValue('Test');
    expect(screen.queryByText(/test/i)).not.toBeInTheDocument();

    fireEvent.keyDown(inputElement, { key: 'Enter' });
    expect(screen.getByText(/test/i)).toBeInTheDocument();
  });

  test('shows active todos', () => {
    render(<App />);

    const activeFilter = screen.getByTestId('active-todos');
    fireEvent.click(activeFilter);
    expect(screen.getByText(/test/i)).toBeInTheDocument();
  });

  test('completes a todo', () => {
    render(<App />);

    const label = screen.getByText(/test/i);
    const checkbox = screen.getByRole('checkbox');

    expect(label).toBeInTheDocument();
    expect(checkbox).not.toBeChecked();

    fireEvent.click(label);
    expect(checkbox).toBeChecked();
  });

  test('clears completed todos', () => {
    render(<App />);

    const clearButton = screen.getByTestId('clear-all');
    expect(screen.getByTestId('list-item')).toBeInTheDocument();

    fireEvent.click(clearButton);
    expect(screen.queryByTestId('list-item')).not.toBeInTheDocument();
  });
});

describe('SNAPSHOT TESTS', () => {
  test('main page should not change', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
});
