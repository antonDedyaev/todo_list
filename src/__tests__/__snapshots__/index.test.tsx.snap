// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAPSHOT TESTS main page should not change 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="container__content"
    >
      <header
        class="container__contentHeader"
      >
        <h1>
          todos
        </h1>
      </header>
      <section
        class="container__contentBody"
      >
        <div
          class="container__inputLine"
        >
          <input
            data-testid="main-input"
            placeholder="What needs to be done?"
            type="text"
            value=""
          />
        </div>
        <div
          class="container__list"
        >
          <div />
        </div>
      </section>
      <footer
        class="container__contentFooter"
      >
        <span>
          0 items left
        </span>
        <div
          class="container__listFilters"
        >
          <button
            class="container__filterButton container__filterButton_active"
            data-testid="display-all"
          >
            All
          </button>
          <button
            class="container__filterButton"
            data-testid="active-todos"
          >
            Active
          </button>
          <button
            class="container__filterButton"
          >
            Completed
          </button>
        </div>
        <button
          data-testid="clear-all"
        >
          Clear completed
        </button>
      </footer>
      <div
        class="container__backPage container__backPage_first"
      />
      <div
        class="container__backPage container__backPage_second"
      />
    </div>
  </div>
</div>
`;
